controllers/machinepool_controller_phases.go

```diff
@@ -0,0 +1,232 @@
+	capierrors "sigs.k8s.io/cluster-api/errors"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+func (r *MachinePoolReconciler) reconcilePhase(mp *clusterv1.MachinePool) {
+	// Set the phase to "pending" if nil.
+	if mp.Status.Phase == "" {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhasePending)
+	}
+
+	// Set the phase to "provisioning" if bootstrap is ready and the infrastructure isn't.
+	if mp.Status.BootstrapReady && !mp.Status.InfrastructureReady {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseProvisioning)
+	}
+
+	// Set the phase to "provisioned" if the infrastructure is ready.
+	if mp.Status.InfrastructureReady {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseProvisioned)
+	}
+
+	// Set the phase to "running" if there is a NodeRef field.
+	if mp.Status.InfrastructureReady && mp.Status.NodeRefs != nil && len(mp.Status.NodeRefs) == int(mp.Status.ReadyReplicas) {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseRunning)
+	}
+
+	// Set the phase to "failed" if any of Status.ErrorReason or Status.ErrorMessage is not-nil.
+	if mp.Status.FailureReason != nil || mp.Status.FailureMessage != nil {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseFailed)
+	}
+
+	// Set the phase to "deleting" if the deletion timestamp is set.
+	if !mp.DeletionTimestamp.IsZero() {
+		mp.Status.SetTypedPhase(clusterv1.MachinePoolPhaseDeleting)
+	}
+}
+
+// reconcileExternal handles generic unstructured objects referenced by a MachinePool
+func (r *MachinePoolReconciler) reconcileExternal(ctx context.Context, mp *clusterv1.MachinePool, ref *corev1.ObjectReference) (*unstructured.Unstructured, error) {
+	obj, err := external.Get(ctx, r.Client, ref, mp.Namespace)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second},
+				"could not find %v %q for MachinePool %q in namespace %q, requeuing",
+				ref.GroupVersionKind(), ref.Name, mp.Name, mp.Namespace)
+		}
+		return nil, err
+	}
+
+	objPatch := client.MergeFrom(obj.DeepCopy())
+
+	// Set external object OwnerReference to the MachinePool.
+	ownerRef := metav1.OwnerReference{
+		APIVersion: clusterv1.GroupVersion.String(),
+		Kind:       "MachinePool",
+		Name:       mp.Name,
+		UID:        mp.UID,
+	}
+
+	if !util.HasOwnerRef(obj.GetOwnerReferences(), ownerRef) {
+		obj.SetOwnerReferences(util.EnsureOwnerRef(obj.GetOwnerReferences(), ownerRef))
+		if err := r.Client.Patch(ctx, obj, objPatch); err != nil {
+			return nil, errors.Wrapf(err,
+				"failed to set OwnerReference on %v %q for MachinePool %q in namespace %q",
+				obj.GroupVersionKind(), ref.Name, mp.Name, mp.Namespace)
+		}
+	}
+
+	// Add watcher for external object, if there isn't one already.
+	_, loaded := r.externalWatchers.LoadOrStore(obj.GroupVersionKind().String(), struct{}{})
```

Comment:
xref #2140 #2168 (no action required now, unless we solve those before this gets merged)